Mayastor-2.0.0-doc update(Phase-1) (#55)
openebs/mayastor-docs: quickstart/configure-mayastor.md

# Configure Mayastor

## Create DiskPool\(s\)

### What is a DiskPool?

> When a node allocates storage capacity for a replica of a persistent volume(PV) it does so from a DiskPool. Each node may create and manage zero, one, or more such pools. The ownership of a pool by a node is exclusive. A pool can manage only one block device, which constitutes the entire data persistence layer for that pool and thus defines its maximum storage capacity.

Reviewer: Should there be a space between "volume" and "(PV)" ?

{% hint style="info" %}
The current version of DiskPools only supports thick provisioning.
{% endhint %}

A pool is defined declaratively, through the creation of a corresponding `DiskPool` custom resource on the cluster. The DiskPool must be created in the same namespace where Mayastor has been deployed.  User configurable parameters of this resource type include a unique name for the pool, the node name on which it will be hosted and a reference to a disk device which is accessible from that node. The pool definition requires the reference to its member block device to adhere to a discrete range of schemas, each associated with a specific access mechanism/transport/ or device type.

> #### Permissible Schemes for `spec.disks` under DiskPool CRD

Reviewer 2: Technically, this is under a DiskPool _CR_ (not CRD).  The CR is what a user creates.  The CRD defines what the expected field structure is.

Author: Will fix this. Thanks.

| Type | Format | Example |
| :--- | :--- | :--- |
| Asynchronous Disk\(AIO\) | Device File | /dev/sdx |

> | Asynchronous Disk I/O \(AIO\) | Device File | aio:///dev/sdx |

Reviewer: Lines 20 and 21 seem the same.

Author: There is just a slight change in the example. The second one has "aio".

| io\_uring | Device File | uring:///dev/sdx |
| Disk(non PCI) with disk-by-guid reference <i>(Best Practice)</i> | Device File | aio:////dev/disk/by-uuid/<uuid_number> OR uring:////dev/disk/by-uuid/<uuid_number> |

Once a node has created a pool it is assumed that it henceforth has exclusive use of the associated block device; it should not be partitioned, formatted, or shared with another application or process. Any pre-existing data on the device will be destroyed.

{% hint style="warning" %}
A RAM drive isn't suitable for use in production as it uses volatile memory for backing the data. The memory for this disk emulation is allocated from the hugepages pool. Make sure to allocate sufficient additional hugepages resource on any storage nodes which will provide this type of storage.
{% endhint %}

### Configure Pools for Use with this Quickstart

> To get started, it is necessary to create and host at least one pool on one of the nodes in the cluster. The number of pools available limits the extent to which the synchronous n-way mirroring feature (replication) of PVs can be configured; the number of pools configured should be equal to or greater than the desired maximum replication factor of the PVs to be created. Also, while placing data replicas ensure that appropriate redundancy is provided. Mayastor's control plane will avoid placing more than one replica of a volume on the same node. For example, the minimum viable configuration for a Mayastor deployment, which is intended to implement 3-way mirrored PVs must have three nodes, each having one DiskPool, with each of those pools having one unique block device allocated to it.

Reviewer 2: There should be no comma after _Mayastor deployment_ and _before which is intended_

It changes the meaning of the sentence.  The Mayastor deployment is the subject of the specific intention.  i.e. the particular mayastor deployment from the example is intended to support 3 way mirrors, so needs three pools.  Another might be intended to support four way, and so would need four pools.  As written here, the minimum viable configuration is always to have three nodes, which is intended to support three way mirrors.

Author: That makes sense. Will update it.

Reviewer: "synchronous N-way mirroring (replication) of PVs"
 Remove "feature" and change n-way to N-way.

We illustrate this quickstart guide with two examples of possible use cases; one which offers no data redundancy \(i.e. a single data replica\), and another having three data replicas. You can modify these as required to match your own desired test cases, within the limitations of the cluster under test.

{% tabs %}
{% tab title="Command \(1 replica example\)" %}
```text
cat <<EOF | kubectl create -f -
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: mayastor-1
parameters:
  ioTimeout: "30"
  protocol: nvmf
  repl: "1"
provisioner: io.openebs.csi-mayastor
reclaimPolicy: Delete

> volumeBindingMode: Immediate

Reviewer 2: I think we should check this with the control plane maintainers.  I'm not sure that immediate binding is the preferred option.

Reviewer 3: I suspect the preference for `WaitForFirstConsumer` was originally intended for the "local" configuration, to place the application, the target/nexus, and the replica on the same node. 
Obviously when n>1 you can't have all replicas on the same node, and if the application moves to a different node then the replica is local no more.
Also with 2.0 we don't need to run the application and target/nexus on the same node..
So unless there's something I'm missing (and I may be), I don't think we need to suggest `WaitForFirstConsumer` from 2.0 onwards, at least for the general use case?

Reviewer: For mayastor 2.0.0, which won't have the nexus HA, shouldn't we stick with "WaitForFirstConsumer" ?

Reviewer 3: What are we trying to achieve by suggesting "WaitForFirstConsumer" ?

Reviewer: Ok. Below is my understanding.
Unless we have high availability for nexus, we won't be able to separate the nexus and application pod.
Since nexus HA is not enabled in 2.0.0, we cannot separate nexus and application pod in 2.0.0, and this is typically enforced by "WaitForFirstConsumer".

Reviewer 3: I don't think we "lock" nexus and application by using "WaitForFirstConsumer". That was done using the "local" flag which has now been removed.
Maybe we should add back the implicitly "local" on the 2.0 until we have HA?
An alternative option which we've discussed previously was to pin the applications to io-engine nodes.
@GlennBullingham @blaisedias

EOF
```
{% endtab %}

{% tab title="Command \(3 replicas example\)" %}
```text
cat <<EOF | kubectl create -f -
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: mayastor-3
parameters:
  ioTimeout: "30"
  protocol: nvmf
  repl: "3"
provisioner: io.openebs.csi-mayastor
reclaimPolicy: Delete
volumeBindingMode: Immediate
EOF
```
{% endtab %}
{% endtabs %}

{% hint style="info" %}

> The default installation of Mayastor includes the deployment of a StorageClass with the name `mayastor-single-replica`. The replication factor is set to 1. The users can either use this default StorageClass or create one themselves.

Reviewer 2: ```suggestion
The default installation of Mayastor includes the creation of a StorageClass with the name `mayastor-single-replica`. The replication factor is set to 1. Users may either use this StorageClass or create their own.
```
We should avoid using the word "default" here, since we can't be certain that the user will have set the SC they create themselves as the default on the cluster.  Nor do we do that for the SC we create on installation.

Author: Okay. Will be updating this.
